phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,459 @@
+        value.getValue().parallelStream().forEach(table -> {
+            try {
+                tableMC.put(table, admin.getDescriptor(TableName.valueOf(table)).isCompactionEnabled());
+                if (!dryRun) {
+                    TableDescriptorBuilder tdb = TableDescriptorBuilder.newBuilder(admin.getDescriptor(TableName.valueOf(table)));
+                    tdb.setCompactionEnabled(false);
+                    admin.modifyTable(tdb.build());
+                }
+                LOGGER.info("Major compaction disabled for "+table);
+
+            } catch (IOException e) {
+                LOGGER.severe("Something went wrong while setting Major Compaction to false for "+table+" "+e);
+            }
+        });
+    }
+
+    private static void setMajorCompaction(Admin admin, Map.Entry<String, ArrayList<String>> value) {
+        value.getValue().parallelStream().forEach(table -> {
+            try {
+                if (!dryRun) {
+                    TableDescriptorBuilder tdb = TableDescriptorBuilder.newBuilder(admin.getDescriptor(TableName.valueOf(table)));
+                    tdb.setCompactionEnabled(tableMC.get(table));
+                    admin.modifyTable(tdb.build());
+                }
+                LOGGER.info("Major compaction set to original value {"+tableMC.get(table)+"} for "+table);
+            } catch (IOException e) {
+                LOGGER.severe("Something went wrong while setting Major Compaction to old value "+table+" "+e);
+            }
+        });
+    }
+
+    private static void rebuildIndexes(Map.Entry<String, ArrayList<String>> value, Configuration conf) {
+        String schema = value.getKey().split("\\.")[0];
+        String table = value.getKey().split("\\.")[1];
+
+        value.getValue().stream().forEach(index -> {
+            IndexTool indexingTool = new IndexTool();
+            conf.set(QueryServices.TRANSACTIONS_ENABLED, Boolean.TRUE.toString());
+            indexingTool.setConf(conf);
+            ArrayList<String> argArray = new ArrayList();
+            argArray.add("-s");
+            argArray.add(schema);
+            argArray.add("-dt");
+            argArray.add(table);
+            argArray.add("-it");
+            argArray.add(index.split("\\.")[1]);
+            // Need to run this job in foreground for the upgrade to be deterministic
+            argArray.add("-direct");
+            argArray.add("-runfg");
+            argArray.add("-op");
+            argArray.add("/tmp/" + UUID.randomUUID().toString());
+            try {
+                LOGGER.info("Rebuilding index "+ index);
+                if(!dryRun) {
+                    indexingTool.run(argArray.toArray(new String[argArray.size()]));
+                }
+            } catch (Exception e) {
+                LOGGER.severe("Something went wrong while building the index "+ index +" " +e);
+            }
+        });
+    }
+
+    private static HashMap<String, ArrayList<String>> getTablesAndIndexes(String tables) {
```

Comment:
Nit: rename to parseTablesAndIndexes or extractTablesAndIndexes

Reply:
done.